using accumulate type to do the computation of norm reduce for cpu path

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92219

As the title, we should use accumulate type to do computation of norm reduce.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10

XiaobingSuper committed:
using accumulate type to do the computation of norm reduce for cpu path

[ghstack-poisoned]

XiaobingSuper committed:
Update on "using accumulate type to do the computation of norm reduce for cpu path"


As the title, we should use accumulate type to do computation of norm reduce.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano commented:
Could be a good time to also have a stab at this one https://github.com/pytorch/pytorch/issues/91597?

ngimel commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -201,8 +201,8 @@ struct AbsMinOps {
 
-  inline C10_DEVICE acc_t project(acc_t a) const {
-    return a;
+  inline C10_DEVICE scalar_t project(acc_t a) const {
```
I think `project` was keeping `acc_t` on purpose, to keep intermediate results that need to be written out in the `acc_t`. 
I don't want to see changes to this file, because it's clearly working for CUDA, and is producing correct results on cuda, so that means that cpu should be able to do cpu-side fixes too

XiaobingSuper committed:
Update on "using accumulate type to do the computation of norm reduce for cpu path"


As the title, we should use accumulate type to do computation of norm reduce.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "using accumulate type to do the computation of norm reduce for cpu path"


As the title, we should use accumulate type to do computation of norm reduce.

cc jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -288,41 +286,41 @@ static void norm_kernel_tensor_iterator_impl(
+          using real_acc_t = at::opmath_type<real_type>;
+          binary_kernel_reduce(
+              iter,
+              AbsMinOps<scalar_t, real_acc_t>(),
```
Even though it is not really necessary for min/max, it should be no harm to do a cast explicitly anyway since we do comparison after upcasting for bf16 or fp16.

jgong5 reviewed (approved):
LGTM

lezcano commented on aten/src/ATen/native/cpu/Reduce.h:
```diff
@@ -118,9 +118,10 @@ static inline void vectorized_outer_reduction(char** data, int64_t inner_stride,
 template<typename traits, typename res_t>
 static void set_result(const int index, const res_t result, const TensorIteratorBase &iter, const int num_outputs) {
   // static_assert(std::is_same<res_t, typename traits::arg2_t>::value, "data types must match");
```
I was going to suggest, shouldn't `out_t` be the same as `res_t`, but I guess not. When does it fail?

lezcano commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -233,23 +233,21 @@ static void norm_kernel_tensor_iterator_impl(
-    });
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+        kHalf, kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
+          using real_type = typename scalar_value_type<scalar_t>::type;
```
Nit. We may want to have an auxiliary
```c++
template<class T>
using scalar_value_t = typename scalar_value_type<T>::value;
```

lezcano commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -233,23 +233,21 @@ static void norm_kernel_tensor_iterator_impl(
+          binary_kernel_reduce(
+              iter, NormOneOps<scalar_t, real_acc_t>(), real_acc_t(0));
+        });
   } else if (val == 2) {
```
Not necessarily in this PR, we may want to extend this to the vectorised case for `val == 2`.

ngimel replied:
We do need to extend this to the vectorized case, after all, this is the most commonly called norm.

lezcano reviewed (commented):
Now we're cooking with gas!
I left a couple comments, but overall it LGTM. Let's wait for Natalia to have a look as well though.

peterbell10 commented on aten/src/ATen/native/cpu/Reduce.h:
```diff
@@ -118,9 +118,10 @@ static inline void vectorized_outer_reduction(char** data, int64_t inner_stride,
 template<typename traits, typename res_t>
 static void set_result(const int index, const res_t result, const TensorIteratorBase &iter, const int num_outputs) {
   // static_assert(std::is_same<res_t, typename traits::arg2_t>::value, "data types must match");
+  using out_t = typename traits::arg2_t;
```
Test failure is real, this breaks `argmax` and `argmin` where the result is not either of the input types. I think `project` really does need to be updated in `SharedReduceOps.h`.

ngimel replied:
Oh so the issue here is that cpu reduce kernel bytecasts the results in the hopes that width is the same and it'll somehow work out (pre-this-pr), and long time ago it had an assert to make sure that it at least makes some sense, but abandoned it long ago? I'd still say that's a very error-prone cpu kernel, and problems should be fixed here.

XiaobingSuper replied:
Yes, I need to rewrite the code as GPU path to fix it.

peterbell10 replied:
> Oh so the issue here is that cpu reduce kernel bytecasts the results in the hopes that width is the same and it'll somehow work out (pre-this-pr)

No, I don't think there's any byte-casting involved. Conceptually, the reduction involves three types: 
- `data_t` - the input type
- `acc_t` - the accumulator type
- `res_t` - the result type

and the type signatures of the reduction op should be:
```
reduce(acc_t, data_t) -> acc_t
combine(acc_t, acc_t) -> acc_t
project(acc_t) -> res_t
```
(This is slightly simplified)

Some of the norm operators have `project(acc_t) -> acc_t` which assumes that `acc_t` is the same as `res_t`. This PR breaks that assumption by accumulating in a higher precision type.

The workaround proposed by this PR is to cast the output of `project` to `data_t`, which forces the output and input dtypes to match. This need not be true in general and in particular is not true for `argmin` and `argmax`.

The correct fix IMO is to change `project` to the correct output dtype in `SharedReduceOps.h` and not touch `Reduce.h`, other than maybe removing the assert that was commented out.

XiaobingSuper replied:
Thanks, but this **project(acc_t) -> out_t** is also used by GPU reduce kernel, I need to check it whether is ok just change it.

XiaobingSuper commented:
I will be on holiday for CNY next two weeks, I will continue to do it when I come back, thanks!
